core/shared/src/main/scala/zio/stm/TReentrantLock.scala

```diff
@@ -0,0 +1,221 @@
+   */
+  lazy val acquireWrite: STM[Nothing, Int] =
+    for {
+      fiberId <- STM.fiberId
+      w <- data.get.collect {
+            case Left(readLock) if (readLock.noOtherHolder(fiberId)) =>
+              WriteLock(1, readLock.readLocks(fiberId), fiberId)
+
+            case Right(WriteLock(n, m, `fiberId`)) =>
+              WriteLock(n + 1, m, fiberId)
+          }
+      _ <- data.set(Right(w))
+    } yield w.writeLocks
+
+  /**
+   * Just a convenience method for applications that only need reentrant locks,
+   * without needing a distinction between readers / writers.
+   *
+   * See [[writeLock]].
+   */
+  final lazy val lock: Managed[Nothing, Int] = writeLock
+
+  /**
+   * Determines if any fiber has a read or write lock.
+   */
+  final def locked: STM[Nothing, Boolean] =
+    (readLocked zipWith writeLocked)(_ || _)
+
+  /**
+   * Obtains a read lock in a managed context.
+   */
+  final lazy val readLock: Managed[Nothing, Int] =
+    Managed.make(acquireRead.commit)(_ => releaseRead.commit)
+
+  /**
+   * Retrieves the number of acquired read locks.
+   */
+  final def readLocks: STM[Nothing, Int] = data.get.map(_.fold(_.readLocks, _.readLocks))
+
+  /**
+   * Determines if any fiber has a read lock.
+   */
+  final def readLocked: STM[Nothing, Boolean] = readLocks.map(_ > 0)
+
+  /**
+   * Releases a read lock held by this fiber. Succeeds with the outstanding
+   * number of read locks held by this fiber.
+   */
+  lazy val releaseRead: STM[Nothing, Int] =
+    STM.fiberId.flatMap(fiberId => adjustRead(fiberId, -1))
+
+  /**
+   * Releases a write lock held by this fiber. Succeeds with the outstanding
+   * number of write locks held by this fiber.
+   */
+  lazy val releaseWrite: STM[Nothing, Int] =
+    STM.fiberId.flatMap(
+      fiberId =>
+        data.modify {
+          case Right(WriteLock(1, m, `fiberId`)) =>
+            0 -> Left(ReadLock(fiberId, m))
+
+          case Right(WriteLock(n, m, `fiberId`)) if n > 1 =>
+            val newCount = n - 1
+
+            newCount -> Right(WriteLock(n - 1, m, fiberId))
+
+          case s => die(s"Defect: Fiber ${fiberId} releasing write lock it does not hold: ${s}")
+        }
+    )
+
+  /**
+   * Obtains a write lock in a managed context.
+   */
+  final lazy val writeLock: Managed[Nothing, Int] =
+    Managed.make(acquireWrite.commit)(_ => releaseWrite.commit)
+
+  /**
+   * Determines if a write lock is held by some fiber.
+   */
+  final def writeLocked: STM[Nothing, Boolean] = writeLocks.map(_ > 0)
+
+  /**
+   * Computes the number of write locks held by fibers.
+   */
+  final def writeLocks: STM[Nothing, Int] = data.get.map(_.fold(_ => 0, _.writeLocks))
+
+  private def adjustRead(fiberId: Fiber.Id, delta: Int): STM[Nothing, Int] =
+    (data.get.collect {
+      case Left(readLock)                    => Left(readLock.adjust(fiberId, delta))
+      case Right(WriteLock(w, r, `fiberId`)) => Right(WriteLock(w, r + delta, fiberId))
```

Comment:
Is it fine here if the number of read locks goes below 0?

Reply:
Fixed by throwing useful die message.